sklearn/utils/_sorting.pyx

```diff
@@ -1,25 +1,121 @@
+cimport numpy as cnp
+
 from cython cimport floating
+from libc.math cimport log as ln
+
+# TODO: In order to support discrete distance metrics, we need to have a
+# simultaneous sort which breaks ties on indices when distances are identical.
+# The best might be using a std::stable_sort and a Comparator which might need
+# an Array of Structures (AoS) instead of the Structure of Arrays (SoA)
+# currently used. Alternatively, we can find a stable algorithm for SoA and
+# adapt it so that it is simultaneous.
+
+# Utilities functions
+
+# TODO: Factor code also present in `tree._utils` or use `libc.math.log2` directly
+cdef inline double log(double x) nogil:
+    return ln(x) / ln(2.0)
+
+def _simultaneous_sort(
+    floating[:] values,
+    cnp.intp_t[:] indices,
+    kind=None,
+):
+    """Interface for testing purposes."""
```

Comment:
Looks better to me. This also makes the Cython code more clear. How does the sorting work if the arrays are multi-dimensional? Worth documenting because it is not clear to me upon skimming?

Some nits: 
- Perhaps also adding a return statement? 
- specify that None -> introvert

Reply:
a1d7194cbd87db68d63b7d73bd799592b8bf3b8d documents it (those routines do not actually return something).